File: Olympiad/COCI/official/2007/contest3_solutions/tenkici.pas

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 3 - Task TENKICI
  Programming language Pascal
*/

program tenkici;

type niz = array[1..500] of integer;

var
   r, s : niz;
   n : integer;
   i, K : longint;
   rjesenje_tenk : array[1..2*500*500] of integer;
   rjesenje_smjer : array[1..2*500*500] of char;

{ rasporedjuje tenkice po retcima (stupcima) }
procedure rasporedi( a : niz; n : integer; X, Y : char );
var
   index : niz;
   i, j, tmp : integer;
begin
   for i := 1 to n do
      index[i] := i;

   { sortiraj }
   for i := 1 to n do
      for j := i+1 to n do
         if a[j] < a[i] then begin
            tmp := a[i]; a[i] := a[j]; a[j] := tmp;
            tmp := index[i]; index[i] := index[j]; index[j] := tmp;
         end;

   { pomakni sve koji idu gore (lijevo) }
   for i := 1 to n do
      while a[i] > i do begin
         a[i] := a[i] - 1;
         K := K + 1;
         rjesenje_tenk[K] := index[i];
         rjesenje_smjer[K] := X;
      end;

   { pomakni sve koji idu dolje (desno) }
   for i := n downto 1 do
      while a[i] < i do begin
         a[i] := a[i] + 1;
         K := K + 1;
         rjesenje_tenk[K] := index[i];
         rjesenje_smjer[K] := Y;
      end;
end;

begin
   readln( n );

   for i := 1 to n do
      readln( r[i], s[i] );

   rasporedi( r, n, 'U', 'D' ); { rasporedi tenkice po retcima }
   rasporedi( s, n, 'L', 'R' ); { rasporedi tenkice po stupcima }

   writeln( K );
   for i := 1 to K do
      writeln( rjesenje_tenk[i], ' ', rjesenje_smjer[i] );
end.

File: Olympiad/COCI/official/2007/contest3_solutions/tenkici.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 3 - Task TENKICI
  Programming language C++
*/

#include <algorithm>
#include <cstdio>
#include <vector>

using namespace std;

typedef pair<int,char> par;
vector<par> rjesenje;

// rasporedjuje tenkice po retcima (stupcima)
void rasporedi( int a[500], int n, char X, char Y ) {
   int index[500];
   for( int i = 0; i < n; ++i ) index[i] = i;

   // sortiraj
   for( int i = 0; i < n; ++i )
      for( int j = i+1; j < n; ++j )
         if( a[j] < a[i] ) {
            swap( a[i], a[j] );
            swap( index[i], index[j] );
         }

   // pomakni sve koji idu gore (lijevo)
   for( int i = 0; i < n; ++i )
      for( ; a[i] > i; --a[i] )
         rjesenje.push_back( par( index[i]+1, X ) );

   // pomakni sve koji idu dolje (desno)
   for( int i = n-1; i >= 0; --i )
      for( ; a[i] < i; ++a[i] )
         rjesenje.push_back( par( index[i]+1, Y ) );
}

int main( void ) {
   int n;
   scanf( "%d", &n );

   int r[500], s[500];
   for( int i = 0; i < n; ++i ) {
      scanf( "%d%d", &r[i], &s[i] );
      --r[i]; --s[i];
   }

   rasporedi( r, n, 'U', 'D' ); // rasporedi tenkice po retcima
   rasporedi( s, n, 'L', 'R' ); // rasporedi tenkice po stupcima

   printf( "%d\n", rjesenje.size() );
   for( vector<par>::iterator it = rjesenje.begin(); it != rjesenje.end(); ++it )
      printf( "%d %c\n", it->first, it->second );

   return 0;
}
